File: 933-number-of-recent-calls/933-number-of-recent-calls.java
import java.util.ArrayList;

class RecentCounter {

  int count;
  ArrayList<Integer> requests;
  int left;
  int right;

  public RecentCounter() {
    count = 0;
    requests = new ArrayList<Integer>();
    left = -1;
    right = -1;
  }

  public int ping(int t) {
    this.requests.add(t);

    if (requests.size() == 1) {
      this.left = 0;
      this.right = 0;
    } else {
      this.right++;
      int min_range = t - 3000;
      int max_range = t;
      int left_element = requests.get(this.left);
      while (left_element < min_range) {
        this.left++;
        left_element = requests.get(this.left);
      }
    }

    return right - left + 1;
  }
}

File: 933-number-of-recent-calls/933-number-of-recent-calls.swift
class RecentCounter {
	var queue: [Int] = []
	var counter: Int = 0
	var min_index: Int = -1
	
	init() {}
	
	func ping(_ t: Int) -> Int {
		if self.queue.count == 0 {
			self.queue.append(t)
			self.counter = 1
			self.min_index = 0
			return 1
		}
		
		self.queue.append(t)
		self.counter += 1
		let min_range: Int = t - 3000
		
		while self.queue[self.min_index] < min_range {
			self.min_index += 1
			self.counter -= 1
		}
		
		return self.counter
	}
}
